SimplenoteUITests/AllNotesClass.swift

```diff
@@ -0,0 +1,93 @@
+//
+//  AllNotesClass.swift
+//  SimplenoteUITests
+//
+//  Created by Sergiy Fedosov on 03.02.2021.
+//  Copyright © 2021 Automattic. All rights reserved.
+//
+
+import XCTest
+
+class AllNotes {
+
+    class func openNote(noteName: String) {
+        app.tables.cells[noteName].tap()
+    }
+    
+    class func isOpen() -> Bool {
+        return app.navigationBars[uidNavBar_AllNotes].exists
+    }
+
+    class func open() {
+        if !isOpen() {
+            app.navigationBars.element.buttons[uidButton_Menu].tap()
+            app.tables.staticTexts[uidButton_AllNotes].tap()
+        }
+    }
+
+    class func addNoteTap() {
+        app.navigationBars[uidNavBar_AllNotes].buttons[uidButton_NewNote].tap()
+    }
+
+    class func createNoteAndLeaveEditor (noteName: String) {
+        AllNotes.addNoteTap()
+        NoteEditor.clearAndEnterText(enteredValue: noteName)
+        NoteEditor.leaveEditor()
+    }
+
+    class func trashNote(noteName: String) {
+        Table.trashNote(noteName: noteName)
+    }
+
+    class func getNotesNumber() -> Int {
+        return Table.getNotesNumber()
+    }
+
+    class func clearAllNotes() {
+        let notesNumber = AllNotes.getNotesNumber()
+        let cellsNum = app.tables.element.children(matching: .cell).count
+        var startingIndex: Int
+
+        if notesNumber == cellsNum {
+            // Depending on what happened before, the cells numbering
+            // might not include "All Notes", "Trash" and "Settings" cells...
+            startingIndex = 0
+        } else {
+            // Or might include them
+            startingIndex = 3
+        }
+
+        for _ in 0..<notesNumber {
+            let cell = app.tables.cells.element(boundBy: startingIndex)
+            cell.swipeLeft()
+            cell.buttons[uidButton_NoteCell_Trash].tap()
+        }
+    }
+
+    class func waitForLoad() {
+        let allNotesNavBar = app.navigationBars[uidNavBar_AllNotes]
+        sleep(4)
+        while allNotesNavBar.staticTexts[uidText_AllNotes_InProgress].exists {
+        }
```

Comment:
Do we need the `sleep` call given there's code that waits already?

Also, I wonder if there's a way to wait for the "in progress" text do disappear that leverages the test framework instead 🤔 This code has the potential of turning into an infinite loop if something breaks in the loading progress.

For example:

```swift
class func waitForLoadedExpectation() -> XCTestExpectation {
    let allNotesNavBar = app.navigationBars[uidNavBar_AllNotes]
    let predicate = NSPredicate { _, _ in
        allNotesNavBar.staticTexts[uidText_AllNotes_InProgress].exists
    }
    return XCTNSPredicateExpectation(predicate: predicate, object: .none)
}
```

Which can then be used like

```swift
wait(for: [AllNotes.waitForLoadedExpectation()], timeout: 30)
```

Reply:
Fixed sleeps and infinite waits are definitely not the way to go. Again, this is something I planned to improve, but kept like that during trial, because of hurry, and forgot about this. Your solution is much better than the one I would probably create: just a finite loop that will stop once the statement is fulfilled or timeout is reached. For the moment I merged the code you proposed into one function, because it seems to me that it will be the only scenario of its usage:

```
    class func waitForLoad() {
        let allNotesNavBar = app.navigationBars[uidNavBar_AllNotes]
        let predicate = NSPredicate { _, _ in
            allNotesNavBar.staticTexts[uidText_AllNotes_InProgress].exists
        }
        
        let expectation = XCTNSPredicateExpectation(predicate: predicate, object: .none)
        XCTWaiter().wait(for: [expectation], timeout: 30)
    }
```

Thank you!